### PR TITLE
deprecates Pubkey::new in favor of Pubkey::{,try_}from

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
`Pubkey::new` lacks type safety.

#### Summary of Changes
The commit deprecates `Pubkey::new` and instead implements `TryFrom<&[u8]>` and `TryFrom<Vec<u8>>` for `Pubkey`.

### CODE REVIEW

File: sdk/program/src/pubkey.rs [hunk 1]
```diff
@@ -121,17 +121,36 @@ impl FromStr for Pubkey {
...
+
+impl TryFrom<Vec<u8>> for Pubkey {
+    type Error = Vec<u8>;
+
+    #[inline]
+    fn try_from(pubkey: Vec<u8>) -> Result<Self, Self::Error> {
+        <[u8; 32]>::try_from(pubkey).map(Self::from)
+    }
+}
+
```
Comment (t-nelson) [inline sdk/program/src/pubkey.rs:153]: can `AsRef<[u8]>` cover both of these?

File: sdk/program/src/pubkey.rs [hunk 2]
```diff
@@ -151,11 +170,12 @@ pub fn bytes_are_curve_point<T: AsRef<[u8]>>(_bytes: T) -> bool {
...
+    #[deprecated(
+        since = "1.14.14",
+        note = "Please use 'Pubkey::from' or 'Pubkey::try_from' instead"
+    )]
     pub fn new(pubkey_vec: &[u8]) -> Self {
-        Self(
-            <[u8; 32]>::try_from(<&[u8]>::clone(&pubkey_vec))
-                .expect("Slice must be the same length as a Pubkey"),
-        )
+        Self(<[u8; 32]>::try_from(pubkey_vec).expect("Slice must be the same length as a Pubkey"))
```
Comment (t-nelson) [inline sdk/program/src/pubkey.rs:178]:
```suggestion
        Self::try_from(pubkey_vec).expect("Slice must be the same length as a Pubkey")
```

dereplicode (i think)

File: sdk/program/src/wasm/pubkey.rs [hunk 3]
```diff
@@ -49,7 +50,8 @@ impl Pubkey {
                 }
                 return Err(format!("Invalid array argument: {:?}", x).into());
             }
-            Ok(Pubkey::new(&bytes))
+            Pubkey::try_from(Pubkey::try_from(bytes))
+                .map_err(|err| Err(format!("Invalid pubkey: {err:?}")))
```
Comment (t-nelson) [inline sdk/program/src/wasm/pubkey.rs:54]: we could improve DX by disambiguating these two error messages.  maybe add the input type?

File: storage-proto/src/convert.rs [hunk 4]
```diff
@@ -496,12 +496,20 @@ impl TryFrom<generated::TransactionStatusMeta> for TransactionStatusMeta {
...
+                    Self::Error::new(bincode::ErrorKind::Custom(err))
+                })?,
             readonly: loaded_readonly_addresses
                 .into_iter()
-                .map(|key| Pubkey::new(&key))
-                .collect(),
+                .map(Pubkey::try_from)
+                .collect::<Result<_, _>>()
+                .map_err(|err| {
+                    let err = format!("Invalid pubkey: {err:?}");
```
Comment (t-nelson) [inline storage-proto/src/convert.rs:510]: ditto

File: rpc/src/rpc.rs [hunk 5]
```diff
@@ -2348,7 +2348,7 @@ fn get_spl_token_owner_filter(program_id: &Pubkey, filters: &[RpcFilterType]) ->
                 ..
             }) if *offset == SPL_TOKEN_ACCOUNT_OWNER_OFFSET => {
                 if bytes.len() == PUBKEY_BYTES {
-                    owner_key = Some(Pubkey::new(bytes));
+                    owner_key = Pubkey::try_from(&bytes[..]).ok();
```
Comment (t-nelson) [inline rpc/src/rpc.rs:2351]: we could eliminate the (now duplicative) length check above with a minor refactor here, right?

File: rpc/src/rpc.rs [hunk 6]
```diff
@@ -2404,7 +2404,7 @@ fn get_spl_token_mint_filter(program_id: &Pubkey, filters: &[RpcFilterType]) ->
                 ..
             }) if *offset == SPL_TOKEN_ACCOUNT_MINT_OFFSET => {
                 if bytes.len() == PUBKEY_BYTES {
-                    mint = Some(Pubkey::new(bytes));
+                    mint = Pubkey::try_from(&bytes[..]).ok();
```
Comment (t-nelson) [inline rpc/src/rpc.rs:2407]: ditto

**Review by t-nelson — COMMENTED** [review]
thanks!

File: sdk/program/src/wasm/pubkey.rs (hunk 3 above)
Comment (behzadnouri) [reply sdk/program/src/wasm/pubkey.rs:54]: what is "DX"?
added input type to error messages.

File: storage-proto/src/convert.rs (hunk 4 above)
Comment (behzadnouri) [reply storage-proto/src/convert.rs:510]: changed the errors to `Invalid {writable,readonly} address`.

File: sdk/program/src/wasm/pubkey.rs (hunk 3 above)
Comment (t-nelson) [reply sdk/program/src/wasm/pubkey.rs]: > what is "DX"?

developer experience

> added input type to error messages.

thanks!

**Review by t-nelson — DISMISSED** [review]
changes lgtm, thanks!

r+ ci

**Pull request merged**